File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int dfs(TreeNode* root, int &ans){
        if(!root) return 0;

        int left=dfs(root->left,ans);
        int right=dfs(root->right,ans);
        left=max(left,0);
        right=max(right,0);
        ans=max(ans,root->val+left+right);

        return root->val+ max(left,right);
    }
    int maxPathSum(TreeNode* root) {
        int ans=INT_MIN;

        dfs(root,ans);

    return ans;}
};

File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        #Every recursive function needs to explicitly return a value in all paths
        def inorderTraversal(node,prev):
            # PATH 1: Empty node
            if not node:
                return True

            # PATH 2: check left subtree. Left subtree fails 
            if not inorderTraversal(node.left,prev):
                return False

            # PATH 3: Current node violates BST
            if node.val <= prev[0]:
                return False

            # Update prev
            prev[0] = node.val

            # PATH 4: Right subtree fails
            if not inorderTraversal(node.right,prev):
                return False

            # PATH 5: ALL checks passed!
            return True
        
        #Python's parameter passing, Mutable types (list, dict, set) 
        # → changes are propagated back
        prev = [float('-inf')]
        valid = inorderTraversal(root,prev)
        return valid
